Initial implementation of ezcap client. (#1)
digitalbazaar/ezcap: README.md

> An easy to use, opinionated Authorization Capabilities (zcap) client library
> for the browser and Node.js.

## Table of Contents

- [Background](#background)
- [Security](#security)
- [Install](#install)
- [Usage](#usage)
- [API Reference](#api-reference)
- [Contribute](#contribute)
- [Commercial Support](#commercial-support)
- [License](#license)

## Background

This library provides a client that browser and node.js applications can use to
interact with HTTP servers protected by zcap-based authorization. The library
is configured with secure and sensible defaults to help developers get started
quickly and ensure that their client code is production-ready.

## Security

The security characteristics of this library are largely influenced by design
that implementers make with respect to private key management. The security
characteristics of the server that you communicate with using this library
are largely dependent on how carefully the server manages zcap registrations,
zcap delegations, and zcap checking. Bugs or failures related to key management,
or enforcement of zcaps will lead to security failures. It is imperative that

> implementers audit their implementations, preferably by parties other than

Reviewer: ```suggestion
implementers audit their implementations, preferably via parties other than
```

Author: Fixed in fa2cf5242ea8b1e4c15feab81cccadccc62ee74e.

To install locally (for development):

```
git clone https://github.com/digitalbazaar/ezcap.git
cd ezcap
npm install
```

## Usage

* [Creating a Client](#creating-a-client)
* [Reading with a Root Capability](#reading-with-a-root-capability)
* [Writing with a Root Capability](#writing-with-a-root-capability)
* [Delegating a Capability](#delegating-a-capability)
* [Reading with a Delegated Capability](#reading-with-a-delegated-capability)
* [Writing with a Delegated Capability](#writing-with-a-delegated-capability)
* [Invoking with a Root Capability](#invoking-a-capability)
* [Invoking with a Delegated Capability](#invoking-a-capability)

### Creating a Client

```js
import {ZcapClient, getCapabilitySigners} from 'ezcap';
import didKey from 'did-method-key';
const didKeyDriver = didKey.driver();

// the base URL for the zcap client to operate against
const baseUrl = 'https://zcap.example';

> // generate a DID Document and set of keypairs

Reviewer: ```suggestion
// generate a DID Document and set of key pairs
```

* [Invoking with a Delegated Capability](#invoking-a-capability)

### Creating a Client

```js
import {ZcapClient} from 'ezcap';
import didKey from 'did-method-key';
const didKeyDriver = didKey.driver();

// the base URL for the zcap client to operate against
const baseUrl = 'https://zcap.example';

// generate a DID Document and set of key pairs
const {didDocument, keyPairs} = await didKeyDriver.generate();

// create a new zcap client using the generated cryptographic material
const zcapClient = new ZcapClient({baseUrl, didDocument, keyPairs});
```

### Reading with a Root Capability

```js
const url = '/my-account/items';

// reading a URL using a zcap will result in an HTTP Response
const response = await zcapClient.read({url});

// retrieve the JSON data
const items = await response.json();
```

### Writing with a Root Capability

```js
const url = '/my-account/items';
const item = {label: 'Widget'};

// writing a URL using a zcap will result in an HTTP Response
const response = await zcapClient.write({url, json: item});

// process the response appropriately
const writtenItem = await response.json();
```

### Delegating a Capability

```js

> const capability = 'https://zcap.example/foo';

Reviewer: ```suggestion
const parentCapability = 'https://zcap.example/foo';
```

Author: Don't agree with this.

const targetDelegate =
  'did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH';
const allowedActions = ['read'];

> const delegatedCapability = zcapClient.delegate({capability, targetDelegate, allowedActions});

Reviewer: Maybe we want to be more explicit that `capability` here is the parent capability.

```suggestion
const delegatedCapability = zcapClient.delegate({parentCapability, targetDelegate, allowedActions});
```

Reviewer: Something that we commonly use to mark and look up zcaps is `referenceId` ... I suspect that we'll find we need that ot be an option to be passed into `delegate` as soon as we start using this library more. JFYI.

Author: Disagree that we want to be more explicit. You have a capability and you want to delegate it. Adding a distinction wrt. 'parentCapability' is something we're trying to hide from people that use ezcap.

Author: Ack wrt. `referenceId` -- but don't want to implement that until we have a concrete place that we're using it.

```

### Reading with a Delegated Capability

```js
const url = '/my-account/items/123';

> const capability = await getCapabilityFromDatabase({url}); // defined by your code

Reviewer: I'm not sure that this will be how capabilities are commonly retrieved; e.g., we often use a reference ID to find zcaps in our applications.

// reading a URL using a zcap will result in an HTTP Response

> const response = await zcapClient.read({url, capability});

Reviewer: There should be no need to pass `url` here, since the capability has an invocation target in it.

Author: Isn't it the case that the `invocationTarget` and where you invoke it /could/ be different... like the `invocationTarget` is a URL like: "https://foo.example/things", but the URL that you invoke it at is "https://foo.example/things/1"?

// retrieve the JSON data
const items = await response.json();
```

### Writing with a Delegated Capability

```js
const url = '/my-account/items';
const item = {label: 'Widget'};
const capability = await getCapabilityFromDatabase({url}); // defined by your code

// writing a URL using a zcap will result in an HTTP Response
const response = await zcapClient.write({url, capability, json: item});

// process the response appropriately
const writtenItem = await response.json();
```

### Invoking with a Root Capability

```js
const url = '/my-account/items';
const item = {count: 12};

// invoking a capability against a URL will result in an HTTP Response
const response = await zcapClient.invoke({url, method: 'patch', json: item});

// process the response appropriately
const updatedItem = await response.json();
```

> ### Invoking with a Delegated Capability

Reviewer: ```suggestion
### Performing a Custom Request with a Delegated Capability
```

const capability = 'https://zcap.example/foo';
const delegate = 'did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH';

> const actions = ['read'];

Reviewer: ```suggestion
const allowedActions = ['read'];
```

> const delegatedCapability = zcapClient.delegate({capability, delegate, actions});

Reviewer: ```suggestion
const delegatedCapability = zcapClient.delegate({capability, delegate, allowedActions});
```

```

### Reading with a Delegated Capability

```js
const url = '/my-account/items/123';
const capability = await getCapabilityFromDatabase({url}); // defined by your code

// reading a URL using a zcap will result in an HTTP Response
const response = await zcapClient.read({url, capability});

// retrieve the JSON data
const items = await response.json();
```

### Writing with a Delegated Capability

```js
const url = '/my-account/items';
const item = {label: 'Widget'};
const capability = await getCapabilityFromDatabase({url}); // defined by your code

// writing a URL using a zcap will result in an HTTP Response
const response = await zcapClient.write({url, capability, json: item});

// process the response appropriately
const writtenItem = await response.json();
```

> ### Invoking with a Root Capability

Reviewer 2: I'm assuming 'invoking' here means - making a general REST request using a zcap that's something other than read or write. In which case, I think it's a somewhat confusing term. Maybe it can be `zcapClient.request()`?

Author: I'm fine w/ the change... @dlongley, thoughts?

Reviewer: Yeah `request` is fine by me.

Author: Done in 9612f0cee6d2886782424d7fc1a6845036d5d926 and e680b0862f9d723d48f5de317a2a9d2c1a466ebe.
